tools/gfx/metal/metal-command-encoder.cpp

```diff
@@ -68,6 +78,62 @@ void ResourceCommandEncoder::copyTexture(
     ITextureResource::Offset3D srcOffset,
     ITextureResource::Extents extent)
 {
+    auto encoder = m_commandBuffer->getMetalBlitCommandEncoder();
+
+    if (dstSubresource.layerCount == 0 && dstSubresource.mipLevelCount == 0 && srcSubresource.layerCount == 0 && srcSubresource.mipLevelCount == 0)
+    {
+        encoder->copyFromTexture(
+            static_cast<TextureResourceImpl*>(src)->m_texture.get(),
+            static_cast<TextureResourceImpl*>(dst)->m_texture.get());
+    }
+    else
+    {
+        for (GfxIndex layer = 0; layer < dstSubresource.layerCount; layer++)
+        {
+            encoder->copyFromTexture(
+                static_cast<TextureResourceImpl*>(src)->m_texture.get(),
+                srcSubresource.baseArrayLayer + layer,
+                srcSubresource.mipLevel,
+                MTL::Origin(srcOffset.x, srcOffset.y, srcOffset.z),
+                MTL::Size(extent.width, extent.height, extent.depth),
+                static_cast<TextureResourceImpl*>(dst)->m_texture.get(),
+                dstSubresource.baseArrayLayer + layer,
+                dstSubresource.mipLevel,
+                MTL::Origin(dstOffset.x, dstOffset.y, dstOffset.z));
+        }
+    }
+}
+
+void ResourceCommandEncoder::copyTextureToBuffer(
+    IBufferResource* dst,
+    Offset dstOffset,
+    Size dstSize,
+    Size dstRowStride,
+    ITextureResource* src,
+    ResourceState srcState,
+    SubresourceRange srcSubresource,
+    ITextureResource::Offset3D srcOffset,
+    ITextureResource::Extents extent)
+{
+    assert(srcSubresource.mipLevelCount <= 1);
+
+    auto encoder = m_commandBuffer->getMetalBlitCommandEncoder();
+    encoder->copyFromTexture(
+        static_cast<TextureResourceImpl*>(src)->m_texture.get(),
+        srcSubresource.baseArrayLayer,
+        srcSubresource.mipLevel,
+        MTL::Origin(srcOffset.x, srcOffset.y, srcOffset.z),
+        MTL::Size(extent.width, extent.height, extent.depth),
+        static_cast<BufferResourceImpl*>(dst)->m_buffer.get(),
+        dstOffset,
+        dstRowStride,
+        dstSize);
+}
+
+void ResourceCommandEncoder::uploadBufferData(
+    IBufferResource* buffer, Offset offset, Size size, void* data)
+{
+    assert(0);
```

Comment:
There is a SLANG_UNIMPLEMENTED macro for this purpose.

Reply:
ok, can do a pass on `assert(0)`

Reply:
done